Backport of xdm PAM setcred security fix for XFree86 4.2.1, taken from
XFree86 4.3.0's stable branch of CVS (xf-4_3-branch).  Backport done by
Mike A. Harris <mharris@redhat.com>.  This backport contains only the
functional changes devoid of the upstream whitespace and formatting
changes for clarity and ease of future maintenance in the earlier 4.x
trees.

--- xc/programs/xdm/session.c.xdm-multiple-security-fixes-CVS11662	2001-12-14 15:01:23.000000000 -0500
+++ xc/programs/xdm/session.c	2003-10-17 10:00:37.000000000 -0400
@@ -537,6 +537,7 @@
 #endif
 #ifdef USE_PAM 
     pam_handle_t *pamh = thepamh();
+    int	pam_error;
 #endif
 
     if (verify->argv) {
@@ -598,8 +599,13 @@
 	}
 #endif   /* QNX4 doesn't support multi-groups, no initgroups() */
 #ifdef USE_PAM
-	if (thepamh()) {
-	    pam_setcred(thepamh(), PAM_ESTABLISH_CRED);
+	if (pamh) {
+	    pam_error = pam_setcred (pamh, PAM_ESTABLISH_CRED);
+	    if (pam_error != PAM_SUCCESS) {
+		LogError ("pam_setcred for \"%s\" failed: %s\n",
+			  name, pam_strerror(pamh, pam_error));
+		return (0);
+	    }
 	}
 #endif
 	if (setuid(verify->uid) < 0)
